File: app/src/main/java/com/worldventures/dreamtrips/wallet/ui/settings/general/reset/FactoryResetDelegate.java
package com.worldventures.dreamtrips.wallet.ui.settings.general.reset;


import com.trello.rxlifecycle.android.RxLifecycleAndroid;
import com.worldventures.dreamtrips.wallet.analytics.WalletAnalyticsAction;
import com.worldventures.dreamtrips.wallet.analytics.WalletAnalyticsCommand;
import com.worldventures.dreamtrips.wallet.analytics.new_smartcard.EnterPinUnAssignAction;
import com.worldventures.dreamtrips.wallet.analytics.new_smartcard.EnterPinUnAssignEnteredAction;
import com.worldventures.dreamtrips.wallet.service.FactoryResetInteractor;
import com.worldventures.dreamtrips.wallet.service.WalletAnalyticsInteractor;
import com.worldventures.dreamtrips.wallet.service.command.FactoryResetCommand;
import com.worldventures.dreamtrips.wallet.service.command.reset.ResetOptions;
import com.worldventures.dreamtrips.wallet.ui.common.navigation.Navigator;

import io.techery.janet.operationsubscriber.OperationActionSubscriber;
import rx.android.schedulers.AndroidSchedulers;

public abstract class FactoryResetDelegate {
   private final FactoryResetInteractor factoryResetInteractor;
   private final WalletAnalyticsInteractor analyticsInteractor;
   private final PinMode pinMode;
   protected final Navigator navigator;

   private FactoryResetDelegate(FactoryResetInteractor factoryResetInteractor,
         WalletAnalyticsInteractor analyticsInteractor, Navigator navigator, PinMode pinMode) {
      this.factoryResetInteractor = factoryResetInteractor;
      this.analyticsInteractor = analyticsInteractor;
      this.navigator = navigator;
      this.pinMode = pinMode;
   }

   public static FactoryResetDelegate create(FactoryResetInteractor factoryResetInteractor,
         WalletAnalyticsInteractor analyticsInteractor, Navigator navigator, FactoryResetAction action) {
      return FactoryResetDelegate.create(factoryResetInteractor, analyticsInteractor, navigator, action, PinMode.ENABLED);
   }

   public static FactoryResetDelegate create(FactoryResetInteractor factoryResetInteractor,
         WalletAnalyticsInteractor analyticsInteractor, Navigator navigator, FactoryResetAction action, PinMode pinMode) {
      if (action == FactoryResetAction.GENERAL) {
         return new GeneralFactoryResetDelegate(factoryResetInteractor, analyticsInteractor, navigator, pinMode);
      } else {
         return new NewCardFactoryResetDelegate(factoryResetInteractor, analyticsInteractor, navigator, pinMode);
      }
   }

   protected final void trackScreen() {
      analyticsInteractor.walletAnalyticsPipe()
            .send(new WalletAnalyticsCommand(trackScreenAnalyticsAction()));
   }

   protected void observeResetSmartCard(FactoryResetView view) {
      factoryResetInteractor.resetSmartCardCommandActionPipe()
            .observe()
            .compose(RxLifecycleAndroid.bindView(view.getView()))
            .observeOn(AndroidSchedulers.mainThread())
            .subscribe(OperationActionSubscriber.forView(view.provideResetOperationView(this))
                  .onSuccess(command -> handleSuccessResult())
                  .create());

      factoryReset();
   }

   public void setupDelegate(FactoryResetView view) {
      if (pinMode == PinMode.ENABLED) {
         performStart();
      } else {
         bindView(view);
      }
   }

   public void factoryReset() {
      factoryResetInteractor.factoryResetCommandActionPipe().send(new FactoryResetCommand(provideResetOptions()));
   }

   public void cancelFactoryReset() {
      factoryResetInteractor.factoryResetCommandActionPipe().cancelLatest();
   }

   public void goBack() {
      navigator.goBack();
   }

   public abstract void bindView(FactoryResetView view);

   protected abstract WalletAnalyticsAction trackScreenAnalyticsAction();

   protected abstract WalletAnalyticsAction trackResetSuccess();

   protected abstract ResetOptions provideResetOptions();

   protected abstract void performStart();

   protected abstract void handleSuccessResult();

   private static class GeneralFactoryResetDelegate extends FactoryResetDelegate {

      private GeneralFactoryResetDelegate(FactoryResetInteractor factoryResetInteractor,
            WalletAnalyticsInteractor analyticsInteractor, Navigator navigator, PinMode pinMode) {
         super(factoryResetInteractor, analyticsInteractor, navigator, pinMode);
      }

      @Override
      protected void handleSuccessResult() {
         navigator.goFactoryResetSuccess();
      }

      @Override
      public void bindView(FactoryResetView view) {
         observeResetSmartCard(view);
      }

      @Override
      protected WalletAnalyticsAction trackScreenAnalyticsAction() {
         throw new UnsupportedOperationException("Factory reset from settings>general doesn't require analytic");
      }

      @Override
      protected WalletAnalyticsAction trackResetSuccess() {
         throw new UnsupportedOperationException("Factory reset from settings>general doesn't require analytic");
      }

      @Override
      protected ResetOptions provideResetOptions() {
         return ResetOptions.builder()
               .withEnterPin(true)
               .build();
      }

      @Override
      protected void performStart() {
         navigator.goFactoryReset();
      }
   }

   private static class NewCardFactoryResetDelegate extends FactoryResetDelegate {

      private NewCardFactoryResetDelegate(FactoryResetInteractor factoryResetInteractor,
            WalletAnalyticsInteractor analyticsInteractor, Navigator navigator, PinMode pinMode) {
         super(factoryResetInteractor, analyticsInteractor, navigator, pinMode);
      }

      @Override
      public void bindView(FactoryResetView view) {
         trackScreen();
         observeResetSmartCard(view);
      }

      @Override
      protected WalletAnalyticsAction trackScreenAnalyticsAction() {
         return new EnterPinUnAssignAction();
      }

      @Override
      protected WalletAnalyticsAction trackResetSuccess() {
         return new EnterPinUnAssignEnteredAction();
      }

      @Override
      protected ResetOptions provideResetOptions() {
         return ResetOptions.builder()
               .withEnterPin(true)
               .wipePaymentCards(false)
               .wipeUserSmartCardData(false)
               .build();
      }

      @Override
      protected void performStart() {
         navigator.goEnterPinUnassign();
      }

      @Override
      protected void handleSuccessResult() {
         trackResetSuccess();
         navigator.goUnassignSuccess();
      }
   }
}

File: wallet/src/main/java/com/worldventures/wallet/ui/settings/general/reset/delegate/FactoryResetDelegate.kt
package com.worldventures.wallet.ui.settings.general.reset.delegate

import com.worldventures.core.janet.composer.ActionPipeCacheWiper
import com.worldventures.wallet.service.FactoryResetInteractor
import com.worldventures.wallet.service.SmartCardInteractor
import com.worldventures.wallet.service.command.FactoryResetCommand
import com.worldventures.wallet.service.command.reset.ResetOptions
import com.worldventures.wallet.ui.common.navigation.Navigator
import io.techery.janet.operationsubscriber.OperationActionSubscriber
import io.techery.janet.smartcard.action.settings.CheckPinStatusAction
import io.techery.janet.smartcard.event.PinStatusEvent
import rx.android.schedulers.AndroidSchedulers

abstract class FactoryResetDelegate internal constructor(
      private val smartCardInteractor: SmartCardInteractor,
      private val factoryResetInteractor: FactoryResetInteractor,
      protected val navigator: Navigator) {

   private var view: FactoryResetView? = null

   @JvmOverloads
   fun factoryReset(resetOptions: ResetOptions = provideResetOptions()) =
         factoryResetInteractor.factoryResetCommandActionPipe().send(FactoryResetCommand(resetOptions))

   fun cancelFactoryReset() = factoryResetInteractor.factoryResetCommandActionPipe().cancelLatest()

   fun goBack() = navigator.goBack()

   fun bindView(view: FactoryResetView) {
      this.view = view
      factoryResetInteractor.resetSmartCardCommandActionPipe()
            .observeWithReplay()
            .compose(view.bindUntilDetach())
            .compose(ActionPipeCacheWiper(factoryResetInteractor.resetSmartCardCommandActionPipe()))
            .observeOn(AndroidSchedulers.mainThread())
            .subscribe(OperationActionSubscriber.forView(view.provideResetOperationView(this))
                  .onSuccess({ this.handleSuccessResult() })
                  .create())
   }

   protected fun checkPinStatus(pinStatusCallback: (Boolean) -> Unit) {
      val localView = view ?: throw IllegalStateException("Method bindView() is not called")
      smartCardInteractor.pinStatusEventPipe()
            .observeSuccess()
            .compose(localView.bindUntilDetach<PinStatusEvent>())
            .observeOn(AndroidSchedulers.mainThread())
            .map({ pinStatusEvent -> pinStatusEvent.pinStatus != PinStatusEvent.PinStatus.DISABLED })
            .take(1)
            .subscribe(pinStatusCallback::invoke)
      smartCardInteractor.checkPinStatusActionPipe().send(CheckPinStatusAction())
   }

   open fun startRegularFactoryReset() = factoryReset()

   protected abstract fun provideResetOptions(): ResetOptions

   protected abstract fun handleSuccessResult()
}

internal open class GeneralFactoryResetDelegate(
      smartCardInteractor: SmartCardInteractor,
      factoryResetInteractor: FactoryResetInteractor,
      navigator: Navigator)
   : FactoryResetDelegate(smartCardInteractor, factoryResetInteractor, navigator) {

   override fun handleSuccessResult() = navigator.goFactoryResetSuccess()

   override fun provideResetOptions(): ResetOptions = ResetOptions.builder()
         .withEnterPin(true)
         .build()
}

internal open class NewCardFactoryResetDelegate(
      smartCardInteractor: SmartCardInteractor,
      factoryResetInteractor: FactoryResetInteractor,
      navigator: Navigator) : FactoryResetDelegate(smartCardInteractor, factoryResetInteractor, navigator) {

   override fun provideResetOptions(): ResetOptions = ResetOptions.builder()
         .withEnterPin(true)
         .wipePaymentCards(false)
         .wipeUserSmartCardData(false)
         .build()

   override fun handleSuccessResult() = navigator.goUnassignSuccess()
}

internal class NewCardFactoryResetDelegateCheckPin(
      smartCardInteractor: SmartCardInteractor,
      factoryResetInteractor: FactoryResetInteractor,
      navigator: Navigator) : NewCardFactoryResetDelegate(smartCardInteractor, factoryResetInteractor, navigator) {

   override fun startRegularFactoryReset() {
      checkPinStatus {
         if (it) {
            performStart()
         } else {
            super.startRegularFactoryReset()
         }
      }
   }

   private fun performStart() = navigator.goEnterPinUnassign()
}

internal class GeneralResetDelegateCheckPin(
      smartCardInteractor: SmartCardInteractor,
      factoryResetInteractor: FactoryResetInteractor,
      navigator: Navigator) : GeneralFactoryResetDelegate(smartCardInteractor, factoryResetInteractor, navigator) {

   override fun startRegularFactoryReset() {
      checkPinStatus {
         if (it) {
            performStart()
         } else {
            super.startRegularFactoryReset()
         }
      }
   }

   private fun performStart() = navigator.goFactoryReset()
}
